### PR TITLE
Improve performance of File.exists?/1

### PR DESCRIPTION
**Pull request by mbuhot**
I've hit a situation where I need to call `File.exists?/1` in the inner loop of a bulk text processing application.

The erlang docs for [file:read_file_info/2](http://erlang.org/doc/man/file.html#read_file_info-2) suggest two optimizations that can speed up calls to this function:

> If the option raw is set, the file server is not called and only information about local files is returned. Note that this will break this module's atomicity guarantees as it can race with a concurrent call to write_file_info/1,2

> As file times are stored in POSIX time on most OS, it is faster to query file information with option posix.

I noticed a significant speedup when calling `read_file_info/2` with these options.
I'm not sure what are the implications of "only information about local files is returned", so maybe that would be a breaking change to `File.exists?/1`.

### CODE REVIEW

**Comment by michalmuskala** [comment]
Yes, the `:raw` option would be a breaking change. What about we make this function accept options, so you could pass them, if that's fine for your use case?

**Comment by michalmuskala** [comment]
Sounds good to me.

**Comment by josevalim** [comment]
I think if we add it here, we should do the same for `dir?` and `regular?` too.

File: lib/elixir/lib/file.ex [hunk 1]
```diff
@@ -178,9 +182,11 @@ defmodule File do
       #=> true
 
   """
-  @spec exists?(Path.t()) :: boolean
-  def exists?(path) do
-    match?({:ok, _}, :file.read_file_info(IO.chardata_to_string(path)))
+  @spec exists?(Path.t(), [exists_option]) :: boolean
+        when exists_option: :raw
+  def exists?(path, opts \\ []) do
+    opts = Keyword.put(opts, :time, :posix)
```
Comment (josevalim) [inline lib/elixir/lib/file.ex]: `[{:time, :posix}] ++ opts` should be enough.

File: lib/elixir/lib/file.ex [hunk 2]
```diff
@@ -166,6 +166,10 @@ defmodule File do
   It can be regular file, directory, socket, symbolic link, named pipe or device file.
   Returns `false` for symbolic links pointing to non-existing targets.
 
+  ## Options
+
+  If the option `:raw` is set, the file server is not called and only information about local files is returned.
```
Comment (josevalim) [inline lib/elixir/lib/file.ex]: By reading this, I would try to set `File.exists?(file, raw: true)` which actually doesn't work. So I think we should say something like:

> The list of options currently support a single option, the atom `:raw`, which
> bypasses the file server and only checks for the file locally.

File: lib/elixir/lib/file.ex [hunk 3]
```diff
@@ -166,6 +166,11 @@ defmodule File do
   It can be regular file, directory, socket, symbolic link, named pipe or device file.
   Returns `false` for symbolic links pointing to non-existing targets.
 
+  ## Options
+
+  The list of options currently support a single option, the atom `:raw`,
```
Comment (fertapric) [inline lib/elixir/lib/file.ex]: I think it should be: "The list of options currently support**s**"

Comment (josevalim) [reply lib/elixir/lib/file.ex]: Thanks @fertapric and sorry for the type @mbuhot! I have pushed a fix!

**Comment by josevalim** [comment]
@mbuhot this looks excellent! Can you please do the same changes to `dir?` and `regular?`? Then we should be good to go. Thank you!

File: lib/elixir/lib/file.ex [hunk 4]
```diff
@@ -166,6 +166,11 @@ defmodule File do
   It can be regular file, directory, socket, symbolic link, named pipe or device file.
   Returns `false` for symbolic links pointing to non-existing targets.
 
+  ## Options
+
+  The list of options currently supports a single option, the atom `:raw`,
+  which bypasses the file server and only checks for the file locally.
```
Comment (ericmj) [inline lib/elixir/lib/file.ex]: Please use a markdown list for the options like we usually do - even though it's just a single option.

**Comment by mbuhot** [comment]
Thanks @josevalim, I've updated `File.dir?` and `File.regular?` similarly.

File: lib/elixir/lib/file.ex [hunk 5]
```diff
@@ -120,15 +120,21 @@ defmodule File do
   This function follows symbolic links, so if a symbolic link points to a
   regular file, `true` is returned.
 
+  ## Options
+
+   - The list of options currently supports a single option, the atom `:raw`,
```
Comment (fertapric) [inline lib/elixir/lib/file.ex]: I think the format @ericmj was suggesting was similar to [this one](https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/file.ex#L382-L387). It's a common style in the Elixir codebase :)

File: lib/elixir/src/elixir_utils.erl [hunk 6]
```diff
@@ -71,7 +71,11 @@ split_last([H], Acc)     -> {lists:reverse(Acc), H};
 split_last([H | T], Acc) -> split_last(T, [H | Acc]).
 
 read_file_type(File) ->
-  case file:read_file_info(File) of
+  read_file_type(File, []).
+
+read_file_type(File, Opts) ->
+  Opts1 = [{time, posix} | Opts],
+  case file:read_file_info(File, Opts1) of
```
Comment (fertapric) [inline lib/elixir/src/elixir_utils.erl]: Minor-minor nitpick: maybe in this case the `Opts1` variable is not required? 
```erlang
case file:read_file_info(File, [{time, posix} | Opts]) of
```

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

File: lib/elixir/lib/file.ex [hunk 7]
```diff
@@ -137,6 +145,13 @@ defmodule File do
   This function follows symbolic links, so if a symbolic link points to a
   directory, `true` is returned.
 
+  ## Options
+
+  The supported options are:
+
+    * `:raw` - a single atom to bypass the file server and only check
+      for the file locally
```
Comment (mbuhot) [inline lib/elixir/lib/file.ex]: should this be 

> for the directory locally

?

Comment (josevalim) [reply lib/elixir/lib/file.ex]: Good catch, i will fix. :)

**Pull request merged**